### DISCUSSION TITLE
Combination of synchronous and asynchronous debounced validators

### DISCUSSION THREAD

**Discussion by xgemelaj** [react-hook-form/react-hook-form#9028] [General]
Because previous ticket (https://github.com/react-hook-form/react-hook-form/issues/888) was closed and there was still discussion about the problem in common case where you are validating new name with synchronous and also with asynchronous validation which is debounced. I am reopening discussion with the @martdavidson text where he described behaviour of this problem.

@bluebill1049 Can you take another look at this?

The _synchronous_ validation returns false, and an error is shown. Then  _asynchronous_ validation resolves and returns true. The error state is now cleared.

Your codesandbox example doesn't have a synchronous validator. Something like:

```
validate: {
      specialCharacters: displayName =>
        meetsAlphaNumericRequirement(displayName) ||
        'The display name can only contain alpha-numeric characters.',
      unique: debounce(checkForUserWithSameDisplayName, 400),
    },
```

`unique` will _always_ resolve after `specialCharacters`, so it doesn't matter what the result of `specialCharacters` was, `unique`'s result will always overwrite it.

In this case, if I enter `Bill$`, `specialCharacters` fails and shows an error, and `unique` is _not_ called with `Bill$` because the criteria is set to `firstErrorFound`. **But** `unique` _will_ fire with just `Bill` at this point, because that was the last value it was called with, and the debounce delay is triggered.

So one path I went down was to call `getValues()` inside of `checkForUserWithSameDisplayName` to get the value of the `displayName` field _at the time that the debounce triggers_ and then call `meetsAlphaNumbericRequirement()` _again_ at the top of that function like so:

```
async function checkForUserWithSameDisplayName(displayName // we actually don't even use this anymore) {
    if (!meetsAlphaNumericRequirement(getValues().displayName)) {
      return 'The display name can only contain alpha-numeric characters.';
    }
   // check uniqueness
}
```

This works, but then the `type` of the error is `unique`, not `specialCharacters`, if you care about that.

Ideally there'd be a way to maybe return undefined from a validator, which would act as though it had never been called, and the previous state of the form would remain.

_Originally posted by @martdavidson in https://github.com/react-hook-form/react-hook-form/issues/888#issuecomment-614283460_

**Comment by bluebill1049** [comment]
what if you combine that validation logic into one which contains both async and sync validation and returns an error message accordingly?

```tsx
validate: (value) => {
  if (Array.isArray(value)) {
     return await validation()
  }
  
  return regularValidation()
}
```

**Reply by xgemelaj** [reply]
I have situation like this: 

`validate={
    required,
    length: length({ min: 5, max: 35 }),
    asyncValidateNewGoalName
}`
              
              
There is problem that synchronous validation should be done first and only in case all synchronous validation are okay I need to do async validation which is debounced. So in this case when async validation is triggered and in debounce time user edit input and sync validations are now incorrect error is shown but once Promise from async validation is resolved then error is overwrite.


I managed to get through this only by doing syncValidation also after i get response from asyncValidation with current value in form from watch('..') ... but it is little bit messy so finding better way to handle this.

**Reply by xgemelaj** [reply]
This is the behaviour if in validate I have syncValidation, asyncValidation

- syncValidtaion -> checking if field is not empty and if length is between 5 and 35 characters
- asyncValidation -> checking uniqueness of name

I have text "TestName" which is correct by both validations. If i start to delete characters by backspace syncValidations are correct until it is 5 characters long so thats the last time asyncValidation is called with debounce with name: "TestN". After that asyncValidation is no more called because syncValidation is returning error. So if i remove more characters and syncValidation is returning error for length of the string and after that asyncValidation is finished for longer string it overwrite syncValidation result.

![ValidationGif](https://user-images.githubusercontent.com/22497924/189905425-30056752-f6d0-45b5-b7b8-05a53541cdce.gif)

**Reply by mdjzon** [reply]
> what if you combine that validation logic into one which contains both async and sync validation and returns an error message accordingly?
> 
> ```tsx
> validate: (value) => {
>   if (Array.isArray(value)) {
>      return await validation()
>   }
>   
>   return regularValidation()
> }
> ```

I have tried to implement our validation basically like you suggest here, but the result from the async `validation()` for the previous value overwrites the result from `regularValidation()` for the latest value. Using version `7.40.0`.

**Reply by mdjzon** [reply]
I'm noticing a difference between using `register()` and `useController()` where the former seem to work as expected but in the case of the latter the async validator overwrites the result of the sync validator:

https://codesandbox.io/s/sweet-darkness-lcsui1

**Reply by mdjzon** [reply]
Would this warrant a new issue, @bluebill1049? Or am I missing something obvious..?

**Reply by bluebill1049** [reply]
can you share some more steps to reproduce the issue? also, the `name` is unique it can't be used twice (in your example), both `register` and `useController` use the same validation mechanism under the hood.

**Reply by bluebill1049** [reply]
in fact, can you reproduce the issue in this CSB: https://codesandbox.io/s/heuristic-brown-0qxrsc?file=/src/App.tsx

**Reply by mdjzon** [reply]
I have replicated the problem: https://codesandbox.io/s/objective-blackwell-o0qhxw?file=/src/App.tsx

There are now two validation functions, one sync that checks for the presence of digits and the async one that isn't particularly fond of periods.

Steps to reproduce:

1. Enter a digit and a period - eg `8.`
2. Remove the period immediately.
3. Observe the error "Please remove period" appear for a brief second, which I believe it should not since it's the late incoming result from the previous validation.

In my first CSB the error remained though until a new validation was triggered so I imagine there is a slight difference in the RHF implementation in `/src`?

**Reply by bluebill1049** [reply]
> Observe the error "Please remove period" appear for a brief second, which I believe it should not since it's the late incoming result from the previous validation.

This is because I subscribe to the `isValid` form state, and the entre form needs to be validated including both async and sync.

**Reply by lmacedotzt** [reply]
The enhancement done on version `7.40.0`

> Improved async validation and especially combined with sync ones, which will abort early sync failed first

Stopped working on version `7.41.1`, the result of the async validation clears the latest error state of the input.

**Reply by kai-dorschner-twinsity** [reply]
This bug is still open and can still reproduced, even with the latest version v7.56.3: https://codesandbox.io/p/sandbox/sharp-black-l2ctgv

There is even a more weird bug:
1. write `8.`
2. remove the period and write it back immediately

👎 What it looks like:
![image](https://github.com/user-attachments/assets/029f7f2a-a0e9-4a0a-a8cd-57134f96424d)

👍 What it **_should_** look like:
![image](https://github.com/user-attachments/assets/ebc4aa43-aa1e-44ad-8e50-527d405d14f8)
